onnxruntime/core/providers/cpu/tensor/gather_elements.cc

```diff
@@ -0,0 +1,276 @@
+
+  const TensorPitches input_shape_pitches(*input_tensor);
+
+  int64_t num_inner_dim = calculate_num_inner_dim(indices_shape);
+  int64_t inner_dim_size = indices_shape[input_rank - 1];
+  bool processing_inner_dim = (axis == input_rank - 1) ? true : false;
+
+  int64_t base_offset = 0;
+  int64_t indices_counter = -1;
+  std::vector<int64_t> process_dims(input_rank, 0);
+
+  size_t element_size = input_tensor->DataType()->Size();
+
+  if (!processing_inner_dim) {
+    while (num_inner_dim-- != 0) {
+      base_offset = compute_base_offset(process_dims, input_shape_pitches, axis);
+
+      // process 1 chunk of 'inner dimension' length
+      for (int64_t i = 0; i < inner_dim_size; ++i) {
+        memcpy(output_data, input_data + (base_offset + (indices_data[++indices_counter] * input_shape_pitches[axis]) + i) * element_size, element_size);
+        output_data += element_size;
+      }
+
+      increment_over_inner_dim(process_dims, indices_shape);
+    }
+  } 
+  // we special-case inner dim as we can weed-out some unnecessary computations in element offset calculations
+  else {
+    while (num_inner_dim-- != 0) {
+      base_offset = compute_base_offset(process_dims, input_shape_pitches, axis);
+
+      // process 1 chunk of 'inner dimension' length
+      for (int64_t i = 0; i < inner_dim_size; ++i) {
+        // for innermost axis, input_shape_pitches[axis] = 1 (so no need to multiply)
+        memcpy(output_data, input_data + (base_offset + indices_data[++indices_counter]) * element_size, element_size);
+        output_data += element_size;
+      }
+
+      increment_over_inner_dim(process_dims, indices_shape);
+    }
+  }
+
+  return Status::OK();
+}
+
+Status GatherElements::Compute(OpKernelContext* context) const {
+  const Tensor* input_tensor = context->Input<Tensor>(0);
+  const TensorShape& input_data_shape = input_tensor->Shape();
+  int64_t input_data_rank = static_cast<int64_t>(input_data_shape.NumDimensions());
+
+  const Tensor* indices_tensor = context->Input<Tensor>(1);
+  const TensorShape& indices_shape = indices_tensor->Shape();
+  int64_t indices_rank = static_cast<int64_t>(indices_shape.NumDimensions());
+
+  if (input_data_rank < 1)
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                           "GatherElements op: Rank of input 'data' needs to be >= 1");
+
+  if (input_data_rank != indices_rank)
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                           "GatherElements op: Rank of input 'data' needs to be equal to rank of input 'indices'");
+
+  Tensor* output_tensor = context->Output(0, TensorShape(indices_shape));
+
+  const auto& input_data_type = input_tensor->DataType();
+  if (input_data_type != output_tensor->DataType())
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                           "GatherElements op: Data type of input 'data' should match the data type of the output");
+
```

Comment:
Shape inferencing validates this so technically this check isn't necessary.

Reply:
Still - I believe it is better to have this check. As a sanity check, we have had it in all places (AFAIK) previously where we do a memcpy between the input and the output.